File: app/src/main/kotlin/com/apex/trader/presentation/screen/positions/PositionsScreen.kt
package com.apex.trader.presentation.screen.positions

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.apex.trader.data.api.dto.PositionDto
import com.apex.trader.presentation.component.ApexCard
import com.apex.trader.presentation.component.KeyValueRow
import com.apex.trader.presentation.component.SidePill
import com.apex.trader.presentation.theme.ApexBear
import com.apex.trader.presentation.theme.ApexBull
import com.apex.trader.presentation.theme.ApexTextMuted

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PositionsScreen(
    onBack: () -> Unit,
    vm: PositionsViewModel = hiltViewModel()
) {
    val ui by vm.ui.collectAsStateWithLifecycle()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Positions & Account") },
                navigationIcon = {
                    IconButton(onClick = onBack) { Icon(Icons.AutoMirrored.Filled.ArrowBack, "Back") }
                },
                actions = {
                    IconButton(onClick = { vm.refresh() }) { Icon(Icons.Default.Refresh, "Refresh") }
                },
                colors = TopAppBarDefaults.topAppBarColors(containerColor = MaterialTheme.colorScheme.background)
            )
        }
    ) { padding ->
        if (ui.loading) {
            Box(modifier = Modifier.fillMaxSize().padding(padding), contentAlignment = Alignment.Center) {
                CircularProgressIndicator()
            }
            return@Scaffold
        }
        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(padding),
            contentPadding = PaddingValues(12.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp)
        ) {
            ui.account?.let { acc ->
                item {
                    ApexCard {
                        Column {
                            Text("Account", style = MaterialTheme.typography.titleMedium)
                            Spacer(Modifier.height(6.dp))
                            KeyValueRow("Wallet balance", "${acc.totalWalletBalance} USDT")
                            KeyValueRow("Margin balance", "${acc.totalMarginBalance} USDT")
                            KeyValueRow("Available", "${acc.availableBalance} USDT")
                            KeyValueRow(
                                "Unrealized PnL",
                                "${acc.totalUnrealizedProfit} USDT",
                                valueColor = if ((acc.totalUnrealizedProfit.toDoubleOrNull() ?: 0.0) >= 0) ApexBull else ApexBear
                            )
                        }
                    }
                }
            }
            if (ui.positions.isEmpty()) {
                item {
                    ApexCard {
                        Text("No open positions", color = ApexTextMuted)
                    }
                }
            } else {
                items(ui.positions, key = { it.symbol + it.positionSide }) { pos ->
                    PositionCard(pos, working = ui.workingSymbol == pos.symbol, onClose = {
                        vm.closePosition(pos.symbol, pos.positionAmt.toDoubleOrNull() ?: 0.0)
                    })
                }
            }
            ui.error?.let {
                item {
                    Text("⚠ $it", color = MaterialTheme.colorScheme.error)
                }
            }
        }
    }
}

@Composable
private fun PositionCard(pos: PositionDto, working: Boolean, onClose: () -> Unit) {
    val amt = pos.positionAmt.toDoubleOrNull() ?: 0.0
    val pnl = pos.pnl.toDoubleOrNull() ?: 0.0
    ApexCard {
        Column {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(pos.symbol, style = MaterialTheme.typography.titleLarge)
                Spacer(Modifier.padding(end = 8.dp))
                SidePill(if (amt >= 0) "LONG" else "SHORT")
                Spacer(Modifier.weight(1f))
                Text(
                    "${pos.leverage}x",
                    color = ApexTextMuted,
                    style = MaterialTheme.typography.labelMedium
                )
            }
            Spacer(Modifier.height(6.dp))
            KeyValueRow("Quantity", "%.6f".format(amt))
            KeyValueRow("Entry", pos.entryPrice)
            pos.markPrice?.let { KeyValueRow("Mark", it) }
            pos.liquidationPrice?.let { KeyValueRow("Liquidation", it, valueColor = ApexBear) }
            KeyValueRow(
                "Unrealized PnL",
                "%.4f USDT".format(pnl),
                valueColor = if (pnl >= 0) ApexBull else ApexBear
            )
            Spacer(Modifier.height(8.dp))
            Button(
                onClick = onClose,
                enabled = !working,
                modifier = Modifier.fillMaxWidth()
            ) { Text(if (working) "Closing…" else "Close at market") }
        }
    }
}

File: lib/features/positions/positions_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/theme.dart';
import '../../data/models/account.dart';
import '../../domain/strategy.dart';
import '../../providers.dart';
import '../../widgets/common.dart';

class PositionsScreen extends ConsumerStatefulWidget {
  const PositionsScreen({super.key});

  @override
  ConsumerState<PositionsScreen> createState() => _PositionsScreenState();
}

class _PositionsScreenState extends ConsumerState<PositionsScreen> {
  bool _loading = true;
  String? _error;
  Account? _account;
  List<Position> _positions = const [];
  String? _workingSymbol;

  @override
  void initState() {
    super.initState();
    _refresh();
  }

  Future<void> _refresh() async {
    setState(() {
      _loading = true;
      _error = null;
    });
    try {
      final repo = ref.read(tradingRepoProvider);
      final account = await repo.getAccount();
      final positions = await repo.getOpenPositions();
      if (!mounted) return;
      setState(() {
        _loading = false;
        _account = account;
        _positions = positions;
      });
    } catch (e) {
      if (!mounted) return;
      setState(() {
        _loading = false;
        _error = e.toString();
      });
    }
  }

  Future<void> _close(Position p) async {
    setState(() => _workingSymbol = p.symbol);
    try {
      final repo = ref.read(tradingRepoProvider);
      final rules = await repo.getSymbolRules(p.symbol);
      if (rules == null) throw Exception('Missing exchange rules for ${p.symbol}');
      await repo.cancelAll(p.symbol);
      await repo.closePosition(
        symbol: p.symbol,
        side: p.isLong ? SignalSide.long : SignalSide.short,
        quantity: p.positionAmt,
        rules: rules,
      );
      if (!mounted) return;
      setState(() => _workingSymbol = null);
      await _refresh();
    } catch (e) {
      if (!mounted) return;
      setState(() {
        _workingSymbol = null;
        _error = e.toString();
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Positions & Account'),
        actions: [
          IconButton(onPressed: _refresh, icon: const Icon(Icons.refresh)),
        ],
      ),
      body: _loading
          ? const Center(child: CircularProgressIndicator())
          : ListView(
              padding: const EdgeInsets.all(12),
              children: [
                if (_account != null)
                  ApexCard(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text('Account', style: Theme.of(context).textTheme.titleMedium),
                        const SizedBox(height: 6),
                        KeyValueRow(
                            label: 'Wallet balance',
                            value: '${_account!.totalWalletBalance.toStringAsFixed(4)} USDT'),
                        KeyValueRow(
                            label: 'Margin balance',
                            value: '${_account!.totalMarginBalance.toStringAsFixed(4)} USDT'),
                        KeyValueRow(
                            label: 'Available',
                            value: '${_account!.availableBalance.toStringAsFixed(4)} USDT'),
                        KeyValueRow(
                          label: 'Unrealized PnL',
                          value: '${_account!.totalUnrealizedProfit.toStringAsFixed(4)} USDT',
                          valueColor: _account!.totalUnrealizedProfit >= 0
                              ? ApexColors.bull
                              : ApexColors.bear,
                        ),
                      ],
                    ),
                  ),
                const SizedBox(height: 10),
                if (_positions.isEmpty)
                  const ApexCard(
                    child: Text('No open positions', style: TextStyle(color: ApexColors.textMuted)),
                  )
                else
                  ..._positions.map(
                    (p) => Padding(
                      padding: const EdgeInsets.only(bottom: 10),
                      child: _positionCard(p),
                    ),
                  ),
                if (_error != null)
                  Padding(
                    padding: const EdgeInsets.only(top: 8),
                    child: Text('⚠ $_error', style: const TextStyle(color: ApexColors.bear)),
                  ),
              ],
            ),
    );
  }

  Widget _positionCard(Position p) {
    // Subscribe to a live mark-price stream for this symbol and compute
    // unrealized P&L on every tick. We use the cached `p.markPrice` /
    // `p.unrealizedProfit` only as the initial value while the first
    // WebSocket frame is in flight.
    final markAsync = ref.watch(markPriceStreamProvider(p.symbol));
    final liveMark = markAsync.maybeWhen(
      data: (t) => t.markPrice,
      orElse: () => p.markPrice,
    );
    final livePnl = liveMark > 0
        ? (liveMark - p.entryPrice) * p.positionAmt
        : p.unrealizedProfit;

    return ApexCard(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            children: [
              Text(p.symbol, style: const TextStyle(fontSize: 18, fontWeight: FontWeight.w700)),
              const SizedBox(width: 8),
              SidePill(side: p.isLong ? 'LONG' : 'SHORT'),
              const Spacer(),
              Text('${p.leverage}x',
                  style: const TextStyle(color: ApexColors.textMuted, fontFamily: 'monospace')),
            ],
          ),
          const SizedBox(height: 6),
          KeyValueRow(label: 'Quantity', value: p.positionAmt.toStringAsFixed(6)),
          KeyValueRow(label: 'Entry', value: p.entryPrice.toStringAsFixed(6)),
          if (liveMark > 0)
            KeyValueRow(label: 'Mark', value: liveMark.toStringAsFixed(6)),
          if (p.liquidationPrice > 0)
            KeyValueRow(
                label: 'Liquidation',
                value: p.liquidationPrice.toStringAsFixed(6),
                valueColor: ApexColors.bear),
          KeyValueRow(
            label: 'Unrealized PnL',
            value: '${livePnl.toStringAsFixed(4)} USDT',
            valueColor: livePnl >= 0 ? ApexColors.bull : ApexColors.bear,
          ),
          const SizedBox(height: 10),
          SizedBox(
            width: double.infinity,
            child: ElevatedButton(
              onPressed: _workingSymbol == p.symbol ? null : () => _close(p),
              child: Text(_workingSymbol == p.symbol ? 'Closing…' : 'Close at market'),
            ),
          ),
        ],
      ),
    );
  }
}
